### PR TITLE
Multiprecision Arithmetic

### PR DESCRIPTION
**Pull request by thetrime**
This is quite a big change. It implements arbitrary[1] precision arithmetic as follows:
- Integers are now represented by IntegerTerm, if possible, or BigIntegerTerm, if not
  - This means that instead of IntegerTerm.get(String image) the parser must now call NumericTerm.get(String image), which returns an appropriate numeric term
- Rational numbers of arbitrary[1] precision are represented by RationalTerm
- You can create a rational using arithmetic functions rdiv/2, rationalize/1 or rational/1
- Availability of this is all determined by vm.Evaluate.isUnbounded. If false, then very large integers parsed from the source turn into floats instead, and the rational functions are not supported (you'll get an evaluation error if you try to use them). The value of this is available via current_prolog_flag(bounded, X)

I've also added 39 tests (of which 18 are skipped if unbounded arithmetic is disabled). These turned up a few existing problems, like (5 >> 64) being given as 5 due to the way Java handles >>. I've fixed those here too.

Finally, I've added another flag vm.Evaluate.strictISO. If this is turned off, then we allow things that ISO doesn't, but make a lot of sense logically speaking. For example, float_fractional_part(3) gives 0, and float_integer_part(3) returns 3, even though ISO says we should raise a type error. Similarly, rounding functions on integers also just echo their input. It also recognises the atom nan as a FloatTerm with value Float.NaN, implements div/2 as integer division that rounds toward -infinity (which is useful since it is guaranteed to be consistent with mod/2), and implements gcd/2 as the GCD of 2 [big]integers

This should probably be documented somewhere, but I'm not entirely sure where. A new section of the manual? Under 'extending'? Is there a sensible way to configure the isUnbounded and/or strictISO flags?  Maybe there's some ant option where we can generate a .java file and include that in the build? 

(Given that you definitely do NOT want to change them during execution, and it seems like it'd be a real effort to provide access to the Environment or Interpreter while /parsing/, I think this is a compile-time setting, not a runtime one)

You might want to bump the version number for this as well (to 0.3.1 or 0.4.0).

1: (subject to available memory, of course)

### CODE REVIEW

**Comment by drt24** [comment]
A new section of the manual yes, probably a new top level section of the manual, something like "Additional features" (though I don't like that name).
It is definitely possible to build a java file or a properties file at ant build time that contains flags like that, I would probably go for a properties file as that makes it easier to change?

File: src/gnu/prolog/term/Rational.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+	public static Rational get(BigInteger val)
+	{
+		return new Rational(val, BigInteger.valueOf(1));
+	}
+
+        private static final double DBL_EPSILON = 0.00000000000000022204;
+        public static Rational getApproximate(double val)
+        {
+                // This is borrowed directly from SWI Prolog
```
Comment (drt24) [inline src/gnu/prolog/term/Rational.java]: Are there any licencing issues with that?

**Comment by drt24** [comment]
This generally looks good. One question on the code, some documentation and some way of configuring it are all that remains to do. If you want documentation and configuration could be in a separate MR.

**Comment by thetrime** [comment]
I'll fix the documentation in this request, but I'm less excited about the configuration. If you have some experience with ant (something I really don't) then I'm happy for some concrete pointers. I spent about an hour trying to work out how to tell ant where to find javacc on my machine but in the end I just had to edit the build.xml and reset it after I'd finished testing, which seems like the /opposite/ of a convenient build system...

For now I'll add it under the "Adding more Prolog functionality" but I propose that (in a separate merge request) I consolidate this, modules and any other non-ISO stuff into a new section called either 'Extensions', 'Non-ISO Extensions' or 'GPJ Extensions' - whichever you prefer.

**Comment by drt24** [comment]
Fair. I learnt some ant for this project because that is what was there when I started but I certainly wouldn't use it for a new project (would probably use maven instead). If someone needs to change the configuration options then they can implement the feature of making that possible.

File: src/gnu/prolog/term/Rational.java (hunk 1 above)
Comment (thetrime) [inline src/gnu/prolog/term/Rational.java]: I'm not sure. The code is inspired very directly by SWI Prolog, but the same algorithm is also implemented in YAP, and apparently in Common LISP. I've found another copy of the paper, at https://www.cs.duke.edu/brd/papers/src-papers/rotations-scg92.pdf and the algorithm is presented in a very similar format in figure 3. I've adjusted the implementation to more closely match the paper as a result.

Comment (drt24) [inline src/gnu/prolog/term/Rational.java]: OK. I think that is probably fine then.

File: docs/manual.texinfo [hunk 2]
```diff
@@ -254,6 +256,109 @@ When compiling a module, shims are installed in user for exported predicate. Sup
 
 The final element that is required for the module system is the fallback to the user module. If the currently executing code tries to call a predicate which is not defined in the current module, it will not decide that the predicate is undefined until it has further checked the user module. Continuing the example above, that means that if you are executing code in a module qux, and call foo(a,b,c) then - assuming there is not a foo/3 also defined in qux, which would take preference - the foo/3 definition in user would be used. This would call :/2, changing the current module to bar, then call foo/3 again, only this time the definition of the exported predicate would be found and executed. The current module is implemented via a stack, and manages restoring the state on backtrack as well as popping modules off the top once the :/2 goal has completed.
 
+@node Multiprecision Arithmetic
+@subsection Large Integers and Rationals
+If gnu.prolog.vm.Evaluate.isUnbounded is set to true then:
```
Comment (drt24) [inline docs/manual.texinfo]: (this is the default)

**Comment by drt24** [comment]
One minor comment on the documentation and then good to go.

**Comment by thetrime** [comment]
Documentation adjusted as requested

**Pull request merged**